modules/standard/Filerator.chpl

```diff
@@ -192,43 +200,114 @@ iter wordexp(pattern="*") {
    By default, it will list all files/directories in the current directory
 */
 
-iter glob(pattern="*") {
+module Chpl_rt_glob {
   extern type glob_t;
   extern proc chpl_glob(pattern:c_string, flags: c_int, 
                         ref ret_glob:glob_t):c_int;
   extern proc chpl_glob_num(x:glob_t): size_t;
   extern proc chpl_glob_index(x:glob_t, idx:size_t): c_string;
   extern proc globfree(ref glb:glob_t);
+}
 
-  var glb : glob_t;
+iter glob(pattern="*") {
+  var glb : Chpl_rt_glob.glob_t;
 
-  const err = chpl_glob(pattern:c_string, 0, glb);
-  const num = chpl_glob_num(glb);
+  const err = Chpl_rt_glob.chpl_glob(pattern:c_string, 0, glb);
+  const num = Chpl_rt_glob.chpl_glob_num(glb);
   if (num) then
     for i in 0..num-1 do
-      yield chpl_glob_index(glb, i): string;
+      yield Chpl_rt_glob.chpl_glob_index(glb, i): string;
+
+  Chpl_rt_glob.globfree(glb);
+}
+
 
+iter glob(pattern:string="*", param tag: iterKind) 
+       where tag == iterKind.standalone {
+  var glb : Chpl_rt_glob.glob_t;
 
-  globfree(glb);
+  const err = Chpl_rt_glob.chpl_glob(pattern:c_string, 0, glb);
+  const num = Chpl_rt_glob.chpl_glob_num(glb);
+  if (num) then
+    forall i in 0..num-1 do
+      yield Chpl_rt_glob.chpl_glob_index(glb, i): string;
+
+  Chpl_rt_glob.globfree(glb);
 }
 
+//
+// TODO: The following leader/follower iterator assumes that
+// the file system state won't change between calls because
+// it calls the glob routines twice and assumes the same
+// result will come back.  In leader-follower 2.0, we would
+// like to store such state between calls in which case this
+// should be rewritten to do so (and would require freeing
+// the state at the end of the call).
+//
+iter glob(pattern:string="*", param tag: iterKind) 
+       where tag == iterKind.leader {
+  var glb : Chpl_rt_glob.glob_t;
+
+  const err = Chpl_rt_glob.chpl_glob(pattern:c_string, 0, glb);
+  //
+  // cast is used here to ensure we create an int-based leader
+  //
+  const num = Chpl_rt_glob.chpl_glob_num(glb).safeCast(int);
+  if (num) {
+    Chpl_rt_glob.globfree(glb);
+
+    //
+    // Forward to the range type's leader
+    //
+    for followThis in (0..num-1).these(tag) do
```

Comment:
I'm sure this is ignorance on my part since the generated code looks right, and I've seen this idiom in DefaultRectangular, but how does this work?

We're serially iterating over an iterator that might be yielding things in parallel? Does this just work because of iterator inlining (is that why we force leaders to be inlined?) Or am I missing something much simpler than that?